features/steps/utils/geometry.py

```diff
@@ -188,23 +188,35 @@ class AlignmentSegmentContinuityCalculation:
     current_start_direction: float = None
     current_start_gradient: float = None
 
-    def _calculate_positions(self) -> None:
+    def _get_u_at_end(self):
+        """
+        Get the value of u corresponding to the end of the previous segment
+        """
+        alignment_rep_type = self.previous_segment.UsingCurves[0].is_a().upper()
+        match alignment_rep_type:
+            case "IFCCOMPOSITECURVE":
+                # Horizontal representation - use the SegmentLength directly
+                u = abs(self.previous_segment.SegmentLength.wrappedValue)
+            case "IFCGRADIENTCURVE" | "IFCSEGMENTEDREFERENCECURVE":
+                # for vertical and cant representations (IfcGradientCurve and IfcSegmentedReferenceCurve),
+                # u should be the total distance traversed by the segment along the base curve
+                # (length of the segment projected to the "Distance Along" x-axis).
+                # The actual distance along the parent curve (e.g. IfcLine, IfcCircle, IfcPolynomialCurve, etc.) will be longer.
+                # Therefore, we will get the projected distance based on the placements of the two segments.
+                u = abs(self.segment_to_analyze.Placement.Location.Coordinates[0] - \
+                    self.previous_segment.Placement.Location.Coordinates[0])
```

Comment:
Isn't this a bit cyclical? We want to assess basically whether segment to analyse and previous segment align, but if we do that based on assessing the respective placements we would miss some cases where the length of previous is actually not enough to reach the placement of current?

In other words: this assesses continuity of vertical in the Y axis, but doesn't account for gaps in the domain of the vertical profile, Is that a problem?

Reply:
Potentially.  I had gotten wrapped around the axle with confusing length along the base curve versus segment length of the parent curve.  Using the placements was the quickest way to resolve this as well as confirm my understanding of the piecewise function implementation.

The cleanest way forward IMHO is to add a property `endPoint` with type `taxonomy::cast<taxonomy::matrix4d>` to [curve_segment_evaluator](https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.8.0/src/ifcgeom/mapping/IfcCurveSegment.cpp#L99) given that we have both `length_` and `projected_length_` available.  Then (I think) we can just call `pwf.end_point()` from python to get the calculated point at the end of the segment - versus dealing with the complication of `length` versus `projected_length` on the python side.

I'm keen to take this on as it will confirm my burgeoning C++ capabilities :wink:.